aptos-move/block-executor/src/limit_processor.rs

```diff
@@ -0,0 +1,206 @@
+};
+use claims::assert_le;
+
+pub struct BlockGasLimitProcessor<T: Transaction> {
+    block_gas_limit_type: BlockGasLimitType,
+    accumulated_effective_block_gas: u64,
+    accumulated_approx_output_size: u64,
+    accumulated_fee_statement: FeeStatement,
+    txn_fee_statements: Vec<FeeStatement>,
+    txn_read_write_summaries: Vec<ReadWriteSummary<T>>,
+}
+
+impl<T: Transaction> BlockGasLimitProcessor<T> {
+    pub fn new(block_gas_limit_type: BlockGasLimitType, init_size: usize) -> Self {
+        Self {
+            block_gas_limit_type,
+            accumulated_effective_block_gas: 0,
+            accumulated_approx_output_size: 0,
+            accumulated_fee_statement: FeeStatement::zero(),
+            txn_fee_statements: Vec::with_capacity(init_size),
+            txn_read_write_summaries: Vec::with_capacity(init_size),
+        }
+    }
+
+    pub(crate) fn accumulate_fee_statement(
+        &mut self,
+        fee_statement: FeeStatement,
+        txn_read_write_summary: Option<ReadWriteSummary<T>>,
+        approx_output_size: Option<u64>,
+    ) {
+        self.accumulated_fee_statement
+            .add_fee_statement(&fee_statement);
+        self.txn_fee_statements.push(fee_statement);
+
+        let conflict_multiplier = if let Some(conflict_overlap_length) =
+            self.block_gas_limit_type.conflict_penalty_window()
+        {
+            self.txn_read_write_summaries.push(
+                if self
+                    .block_gas_limit_type
+                    .use_granular_resource_group_conflicts()
+                {
+                    txn_read_write_summary.unwrap()
+                } else {
+                    txn_read_write_summary
+                        .unwrap()
+                        .collapse_resource_group_conflicts()
+                },
+            );
+            self.compute_conflict_multiplier(conflict_overlap_length as usize)
+        } else {
+            1
+        };
+
+        // println!("[{}] Multiplier {}, with read/write summary: {:?}", self.txn_fee_statements.len() - 1, conflict_multiplier, self.txn_read_write_summaries.last());
+
+        // When the accumulated execution and io gas of the committed txns exceeds
+        // PER_BLOCK_GAS_LIMIT, early halt BlockSTM. Storage fee does not count towards
+        // the per block gas limit, as we measure execution related cost here.
+        self.accumulated_effective_block_gas += conflict_multiplier
+            * (fee_statement.execution_gas_used()
+                * self
+                    .block_gas_limit_type
+                    .execution_gas_effective_multiplier()
+                + fee_statement.io_gas_used()
+                    * self.block_gas_limit_type.io_gas_effective_multiplier());
+
+        self.accumulated_approx_output_size += approx_output_size.unwrap_or(0);
+    }
+
+    fn should_end_block(&self, is_parallel: bool) -> bool {
+        let mode = if is_parallel {
```

Comment:
supernit: could we just pass this directly? (I suppose we call from either sequential or parallel context and can directly provide the proper one?)

Reply:
this way it is explicit on the caller (vs boolean true/false whether it means sequential or parallel